### ISSUE TITLE
Prohibit stackalloc int[3][1]

### ISSUE THREAD

**Issue by jskeet** [dotnet/csharpstandard#1232]
In #1211 it looks like `stackalloc int[3][1]` should be valid by the standard, but Roslyn complains. We've decided it wouldn't do what is probably expected anyway, so it would be better to prohibit it in the standard.

We haven't yet decided on the best approach for this (e.g. in the grammar or as an additional semantic rule in the text).

**Comment by gafter** [comment]
Possible strategy: Categorize it as an array creation expression in the grammar.

**Comment by jnm2** [comment]
Would it help those who come after us to say something explicit, so that people don't have to try to tease apart whether or not things were blocked by design and what the rationale might have been?

**Comment by gafter** [comment]
@jnm2 Yes, and the proposed solution piggy-backs on such text already in the standard:

> Primary expressions are divided between *array_creation_expression*s and *primary_no_array_creation_expression*s. Treating *array_creation_expression* in this way, rather than listing it along with the other simple expression forms, enables the grammar to disallow potentially confusing code such as
> 
> ```csharp
> object o = new int[3][1];
> ```
> 
> which would otherwise be interpreted as
> 
> ```csharp
> object o = (new int[3])[1];
> ```

**Comment by jskeet** [comment]
We need to see whether https://github.com/dotnet/csharpstandard/pull/1322 solved this. We're not sure.

**Comment by jskeet** [comment]
Assigned @gafter just to check if this is still an issue.

**Comment by gafter** [comment]
I don't think so. This is still an issue, and worse than before. #1322 has been merged, but in https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/grammar.md we have the following:

```antlr
primary_expression
    : stackalloc_expression
    | element_access
    | array_creation_expression
    | ...
    ;
element_access
    : primary_expression '[' argument_list ']'
    ;
```

In short, it appears that the parts of the spec that would have forbidden `new int[3][1]` have been removed from the specification, and cannot be relied on to also forbid `stackalloc int[3][1]`.

**Comment by Nigel-Ecma** [comment]
> I don't think so. This is still an issue, and worse than before. [#1322](https://github.com/dotnet/csharpstandard/pull/1322) has been merged, but in https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/grammar.md we have the following:
> 
> primary_expression
>     : stackalloc_expression
>     | element_access
>     | array_creation_expression
>     | ...
>     ;
> element_access
>     : primary_expression '[' argument_list ']'
>     ;
> In short, it appears that the parts of the spec that would have forbidden `new int[3][1]` have been removed from the specification, and cannot be relied on to also forbid `stackalloc int[3][1]`.

@gafter – #1322 switched from syntactic restriction to semantic restrictions, e.g. in for *element_access* it now says:

> The *primary_expression* of an *element_access* may not be an *array_creation_expression* unless it includes an *array_initializer*, or a *stackalloc_expression* unless it includes a *stackalloc_initializer*.

Followed by a *Note* explaining this prevent these forms. The preamble in #1322 explains why this was done.

Hopefully the restrictions are all there and correct.

**Comment by jskeet** [comment]
@gafter: Please could you check again with the context of @Nigel-Ecma's comment?

**Comment by BillWagner** [comment]
Notes for the March meeting:

I propose that we close this without further action.

https://github.com/dotnet/csharpstandard/issues/1232#issuecomment-3134421170 is correct: The restriction is now in the prose directly following the grammar for both *Element access* and *Null conditional element access*.

Further, the grammar for a stackalloc expression prohibits a multi-dimensional stackalloc expression.

I don't see further changes needed.

One alternative would be to add text in 6.5 that describes this syntactic ambiguity and how it is resolved in prose.

**Comment by Nigel-Ecma** [comment]
This can be closed. The topic was addressed in PR #1322, indeed that went further to allow a stackalloc (and array creation) with an initialiser to be directly indexed which the original formulation missed.

I think part of the confusion here is the tendency, at least by me, to collectively refer to rules in prose as “semantic rules” when they can be syntactic restrictions, as here in for *element_access*, syntax disambiguations, etc. There are ”semantic” as they are not “syntactic”, and also often translate to ANTLR semantic predicates or actions.

**Comment by jskeet** [comment]
Closed as implemented in #1322.
